### ISSUE TITLE
Should we make relocations specific to data segments and individual functions

### ISSUE THREAD

**Issue by sbc100** [WebAssembly/tool-conventions#32]
By default in the toolchain we are running with -ffunction-sections and -fdata-sections.  In order to achieve this the linker must be able to selectively include or exclude a given data segment or an individual function.

However, relocations are specified on the per-section basis, which means that figuring out which relocations apply to given segment or function is not easy (currently all relocations are scanned for each segment/function).

Rather than one reloc section per input section (i.e. reloc.CODE + reloc.DATA) should we include a reloc section for each segment/function.  e.g. reloc.CODE.1, reloc.CODE.2, reloc.DATA.1, etc...

**Comment by dschuff** [comment]
Since these are custom sections,  I think it makes sense to do whatever is easier/faster for the linker. So if I understand you correctly, the linker traverses the functions, and then for each function has to scan all the relocs to find the ones that apply to it? Are the relocs ordered by function in the object file, or can they be made that way? i.e. could the linker just keep track of where it left off from the previous function when starting the next? (or even if that's difficult in the implementation, I guess we could keep an index of where the first relocation for each function is... but maybe that's no better than just having one section per function).

**Comment by NWilson** [comment]
We now have just the one place we do this (I think), in `InputChunk::copyRelocations`, where we create for each function/segment a list of the relocations that apply to it. If were to simply iterate in sorted order it must be possible to do it all in a single pass.

I'm not saying it's bad to change the relocation data, there might be other reasons for doing that, but the performance issue could be solved without changing the file format. In the very worst case, building a temporary map would do the job, and we do that in other places too while parsing.

**Comment by NWilson** [comment]
The optimisation to do a single pass (rather than n^2 traversal) is under review here: https://reviews.llvm.org/D42176

**Comment by sbc100** [comment]
I think its more than just the performance issue.  The fundamental units the the wasm linker is working with are segments and functions, not sections, so I still think there is an argument for being explicit about which function or segment the relocations apply to.  

We have a somewhat similar problem with symbols.   Each symbol should really know which segment it is in.   We have a hack in llvm today that prevents a zero size symbol pointing the end of a segment (we just add one to it (see LastFragmentSize in WasmObjectWriter.cpp).  So I think symbols should be segment specific too, like they are in ELF.

**Comment by NWilson** [comment]
Sam, how about having one reloc section still (one section for functions, one for code), but we could add an explicit "subsection index" to each relocation:

```
  struct WasmRelocation {
+   uint32_t ElementIndex; // Index of the function/segment the reloc modifies.
    uint8_t Type;    // The type of the relocation.
    uint32_t Index;  // Index into either symbol or type index space.
    uint64_t Offset; // Offset from the start of the section.
    int64_t Addend;  // A value to add to the symbol.
  };
```

Then you wouldn't have to compute the function by working out from the offset, and the offset could be validated to lie within the specified function/segment. That would be fairly neat, and would avoid having a whole load of relocation custom sections. 

Your idea of using "reloc.CODE.N" goes against the current design where the name after "reloc." is ignored (not that I'm 100% convinced by the design).

On the other hand, adding sub-indexes inside the reloc section also goes against the current design a bit, in that relocations are currently supposed to be "generic" ie without any knowledge of data or function sections. Currently you're able to apply relocs for custom sections you don't know anything about, but requiring the linker to know what's inside the section is a bit of a shame.

I'm still fairly happy with the status quo, in that the offset does efficiently and uniquely identify the function/segment for each reloc. I'm just trying to suggest an alternative or more concrete suggestion for how we'd actually encode the function/segment index in the reloc section.

**Comment by sbc100** [comment]
I was considering basically the same thing.  I was going to call it `Chunk`, and I was thinking we could group relocations within a reloc section.    So we don't need to repeat the `Chunk` for each reloc.

**Comment by sbc100** [comment]
The only wrinkle is that we might want to handle sections that are not split into chunks too.   Perhaps Chunk 0 can be used in this case?

**Comment by dschuff** [comment]
Yeah, whatever we come up with has to work for debug info too. Is the idea still to have one reloc section for each of code and data, and one for each custom section that needs relocations applied? We could just have those be 1 chunk, or perhaps we might even end up using subsections there too.
